number_theory/numeric.h

```diff
@@ -92,10 +95,59 @@ auto pow(T base, U exponent) -> decltype(std::pow(base, exponent)) {
   return std::pow(base, exponent);
 }
 
+// Returns the |n|-th root of |x| rounded to integer toward zero.
+// The argument |n| shoud be positive.
+// Throws domain_error exception if the root does not exist.
+// The time complexity is O(log |x|).
+//
+// This overload is only enabled for x < 2^64, because it uses uint64_t
+// to avoid overflows during the computation.
+template <typename T,
+          typename U,
+          std::enable_if_t<std::numeric_limits<T>::digits <= 64, bool> = true>
+T iroot(const T &x, const U &n) {
+  static_assert(
+      std::numeric_limits<T>::is_integer && std::numeric_limits<U>::is_integer,
+      "iroot arguments must be integers.");
+  if (n < 0)
+    throw std::invalid_argument("iroot argument |n| should be positive.");
+  if (n == 0)
+    throw std::domain_error("The 0-th root does not exist.");
+  if (x < 0 && n % 2 == 0)
+    throw std::domain_error("Negative numbers do not have even roots.");
+
+  if (n == 1)
+    return x;
+
+  // For each n, the table stores the maximum y that y^n does not overflow.
+  static const int max_size = 64;
+  static const uint64_t y_limits[max_size] = {
+      0,  0,  4294967295, 2642245, 65535, 7131, 1625, 565, 255, 138, 84, 56, 40,
+      30, 23, 19,         15,      13,    11,   10,   9,   8,   7,   6,  6,  5,
+      5,  5,  4,          4,       4,     4,    3,    3,   3,   3,   3,  3,  3,
+      3,  3,  2,          2,       2,     2,    2,    2,   2,   2,   2,  2,  2,
+      2,  2,  2,          2,       2,     2,    2,    2,   2,   2,   2,  2};
+  uint64_t y_limit = (n >= max_size ? 1 : y_limits[n]);
```

Comment:
I am not sure if the surrounding parentheses contributes to readability here. It might even look cleaner without them. What do you think?